programs/bpf_loader/src/serialization.rs

```diff
@@ -167,18 +167,25 @@ pub fn deserialize_parameters_unaligned(
             start += size_of::<u8>(); // is_signer
             start += size_of::<u8>(); // is_writable
             start += size_of::<Pubkey>(); // key
-            let _ = borrowed_account.set_lamports(LittleEndian::read_u64(
+            let lamports = LittleEndian::read_u64(
                 buffer
                     .get(start..)
                     .ok_or(InstructionError::InvalidArgument)?,
-            ));
+            );
+            if borrowed_account.get_lamports() != lamports {
+                borrowed_account.set_lamports(lamports)?;
+            }
             start += size_of::<u64>() // lamports
                 + size_of::<u64>(); // data length
-            let _ = borrowed_account.set_data(
-                buffer
-                    .get(start..start + pre_len)
-                    .ok_or(InstructionError::InvalidArgument)?,
-            );
+            let data = buffer
+                .get(start..start + pre_len)
+                .ok_or(InstructionError::InvalidArgument)?;
+            // The redundant `is_data_mutable` helps to avoid the expensive data comparison if we can
+            if borrowed_account.is_data_mutable(data.len()).is_ok()
```

Comment:
When `enable_early_verification_of_account_modifications` is true, does this mean that if the account is not mutable we don't check the data?  Is that because if not mutable the program fails immediately if it attempts to write to it?

Reply:
For the runtime and built-in programs that would be the case yes, but this here is the deserialization edge (same goes for the CPI syscall) of BPF programs with ABIv0 or ABIv1. And once `enable_early_verification_of_account_modifications` is active, the following will be the case:
- BPF programs with ABIv0 or ABIv1 will continue to experience their errors at the Instruction or CPI edges, but here in the deserialization and the CPI syscall, instead of in `PreAccount::verify()`. So it is still "earlier" in that sense.
- Accounts for which `is_data_mutable() == false` will check if their data changed, and if so try to attempt to modify it, which will result in the correct Error being thrown. That will cost one data comparison, instead of the copy and comparison now.
- Accounts for which `is_data_mutable() == true` will not check if their data changed, and directly proceed to modify / overwrite it. That will cost one data copy, instead of the copy and comparison now.